autoload/lsp/internal/work_done_progress.vim

```diff
@@ -0,0 +1,62 @@
+" https://microsoft.github.io/language-server-protocol/specifications/specification-current/#progress
+
+let s:progress_ui = []
+let s:enabled = 0
+
+function! lsp#internal#work_done_progress#_enable() abort
+    if !g:lsp_work_done_progress_enabled | return | endif
+
+    if s:enabled | return | endif
+    let s:enabled = 1
+
+    let s:Dispose = lsp#callbag#pipe(
+          \ lsp#stream(),
+          \ lsp#callbag#filter({x->has_key(x, 'response') && has_key(x['response'], 'method')
+          \  && x['response']['method'] ==# '$/progress' && has_key(x['response'], 'params')
+          \  && has_key(x['response']['params'], 'value') && type(x['response']['params']['value']) == type({})}),
+          \  lsp#callbag#subscribe({'next': {x->s:handle_work_done_progress(x['server'], x['response'])}})
+          \ )
+endfunction
+
+function! s:handle_work_done_progress(server, progress) abort
+    let l:value = a:progress['params']['value']
+    " Add the server name to distinguish the server
+    let l:token = a:server . ':' . a:progress['params']['token']
+    let l:new = {
+      \ 'server': a:server,
+      \ 'token': l:token,
+      \ 'title': '',
+      \ 'messages': '',
+      \ 'percentage': -1,
+      \ }
+
+    if l:value['kind'] ==# 'end'
+        let l:new['messages'] = ''
+        let l:new['percentage'] = 100.0
+        let s:progress_ui = filter(s:progress_ui, {_, x->x['token'] !=# l:token})
+    elseif l:value['kind'] ==# 'begin'
+        let l:new['title'] = l:value['title']
+        let s:progress_ui = filter(s:progress_ui, {_, x->x['token'] !=# l:token})->insert(l:new)
+    elseif l:value['kind'] ==# 'report'
+        let l:new['messages'] = get(l:value, 'message', '')
+        let l:new['percentage'] = get(l:value, 'percentage', -1.0)
```

Comment:
spec says percentage should be uinteger.

```vim
if has_key(l:value, 'percentage') | let l:new['percentage'] = l:value['percentage'] | endif

Reply:
I'm sorry.
I misunderstood the specification, because clangd and rust-analyzer return float.
But they should be 'uinteger' in the future.
microsoft/language-server-protocol#1037

So I fixed to uinteger.